x-pack/plugins/triggers_actions_ui/public/application/sections/alerts_table/hooks/use_columns.ts

```diff
@@ -0,0 +1,228 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { EuiDataGridColumn } from '@elastic/eui';
+import { IStorageWrapper } from '@kbn/kibana-utils-plugin/public';
+import { AlertConsumers } from '@kbn/rule-data-utils';
+import { BrowserField, BrowserFields } from '@kbn/rule-registry-plugin/common';
+import { useCallback, useEffect, useState } from 'react';
+import { AlertsTableStorage } from '../alerts_table_state';
+import { useFetchBrowserFieldCapabilities } from './use_fetch_browser_fields_capabilities';
+
+interface UseColumnsArgs {
+  featureIds: AlertConsumers[];
+  storageAlertsTable: React.MutableRefObject<AlertsTableStorage>;
+  storage: React.MutableRefObject<IStorageWrapper>;
+  id: string;
+  defaultColumns: EuiDataGridColumn[];
+}
+
+const fieldTypeToDataGridColumnTypeMapper = (fieldType: string | undefined) => {
+  if (fieldType === 'date') return 'datetime';
+  if (fieldType === 'number') return 'numeric';
+  if (fieldType === 'object') return 'json';
+  return fieldType;
+};
+
+/**
+ * EUI Data Grid expects the columns to have a property 'schema' defined for proper sorting
+ * this schema as its own types as can be check out in the docs so we add it here manually
+ * https://eui.elastic.co/#/tabular-content/data-grid-schema-columns
+ */
+const euiColumnFactory = (
+  column: EuiDataGridColumn,
+  browserFields: BrowserFields
+): EuiDataGridColumn => {
+  const browserFieldsProps = getBrowserFieldProps(column.id, browserFields);
+  return {
+    ...column,
+    schema: fieldTypeToDataGridColumnTypeMapper(browserFieldsProps.type),
+  };
+};
+
+/**
+ * Searches in browser fields object for a specific field
+ */
+const getBrowserFieldProps = (
+  _columnId: string,
```

Comment:
why this parameter starts with `_`? just curious

Reply:
There is this error that columnId is already defined in an upper scope. How do you usually handle that type of error?